src/tools/database/elasticsearch.ts

```diff
@@ -0,0 +1,209 @@
+    });
+  }
+
+  static {
+    this.register();
+  }
+
+  public constructor(options: ElasticSearchToolOptions) {
+    super(options);
+    if (!options.connection.cloud && !options.connection.node && !options.connection.nodes) {
+      throw new ValidationError([
+        {
+          message: "At least one of the properties must be provided",
+          propertyName: "connection.cloud, connection.node, connection.nodes",
+        },
+      ]);
+    }
+  }
+
+  @Cache()
+  protected async client(): Promise<Client> {
+    try {
+      const client = new Client(this.options.connection);
+      await client.info();
+      return client;
+    } catch (error) {
+      throw new ToolError(`Unable to connect to ElasticSearch: ${error}`, [], {
+        isRetryable: false,
+        isFatal: true,
+      });
+    }
+  }
+
+  protected async _run(
+    input: ToolInput<this>,
+    _options: BaseToolRunOptions | undefined,
+    run: RunContext<this>,
+  ): Promise<JSONToolOutput<any>> {
+    if (input.action === ElasticSearchAction.ListIndices) {
+      const indices = await this.listIndices(run.signal);
+      return new JSONToolOutput(indices);
+    } else if (input.action === ElasticSearchAction.getIndexDetails) {
+      const indexDetails = await this.getIndexDetails(input, run.signal);
+      return new JSONToolOutput(indexDetails);
+    } else if (input.action === ElasticSearchAction.search) {
+      const response = await this.search(input, run.signal);
+      if (response.aggregations) {
+        return new JSONToolOutput(response.aggregations);
+      } else {
+        return new JSONToolOutput(response.hits.hits.map((hit: SearchHit) => hit._source));
+      }
+    } else {
+      throw new ToolError(`Invalid action specified: ${input.action}`);
+    }
+  }
+
+  private async listIndices(signal?: AbortSignal): Promise<CatIndicesResponse> {
+    const client = await this.client();
+    const response = await client.cat.indices(
+      {
+        expand_wildcards: "open",
+        h: "index",
+        format: "json",
+      },
+      { signal: signal },
+    );
+    return response
+      .filter((record) => record.index && !record.index.startsWith(".")) // Exclude system indices
+      .map((record) => ({ index: record.index }));
+  }
+
+  protected async getIndexDetails(
+    input: ToolInput<this>,
+    signal: AbortSignal,
+  ): Promise<IndicesGetMappingResponse> {
+    if (!input.indexName) {
+      throw new ToolError("Index name is required for GET_INDEX_DETAILS action.");
+    }
+    const client = await this.client();
+    return await client.indices.getMapping(
+      {
+        index: input.indexName,
+      },
+      { signal: signal },
+    );
+  }
+
+  protected async search(input: ToolInput<this>, signal: AbortSignal): Promise<SearchResponse> {
+    if (!input.indexName || !input.query) {
+      throw new ToolError("Both index name and query are required for SEARCH action.");
```

Comment:
This is more like `ToolInputValidationError`.

Also, it should be ideal to do the whole validation in the `validateInput` method so that once the tool is called, it is called with the correct parameters. The next advantage is that if you throw the `ToolInputValidationError` error, it gets propagated back to the agent so they can recover.